h2/src/main/org/h2/result/LocalResult.java

```diff
@@ -417,10 +382,34 @@ private void applyOffsetAndLimit() {
             rows = new ArrayList<>(rows.subList(offset, offset + limit));
         } else {
             if (clearAll) {
+                external.close();
+                external = null;
                 return;
             }
-            diskOffset = offset;
+            trimExternal(offset, limit);
+        }
+    }
+
+    private void trimExternal(int offset, int limit) {
+        ResultExternal temp = external;
+        external = null;
+        temp.reset();
+        while (--offset >= 0) {
+            temp.next();
+        }
```

Comment:
Just an idea for an optimisation - maybe add a skip(int) method to ResultExternal and use it here?